### PR TITLE
TreeDB Raft: route Mongo gateway writes through cluster submitter

### LINKED ISSUE #3258: TreeDB Raft #3044-3: Mongo gateway cluster submitter adapter
Parent: #3044
Depends on: #3247 / PR #3249
Conditional dependency: #3246 / PR #3254 for `update_bson_set`

## Goal

Route supported Mongo gateway writes through the shared deterministic cluster submitter seam instead of mutating local collections directly when cluster mode is configured.

## Scope

- [ ] Add Mongo gateway server configuration for an optional `nativewire.ClusterSubmitter`.
- [ ] Thread request context to Mongo command handlers so submitter calls can honor connection/request cancellation.
- [ ] Build deterministic native-wire command entries from Mongo commands using `TreeDB/internal/nativewire`.
- [ ] Route supported explicit `create`, `insert`, `delete`, and, if #3246 is present, top-level BSON `$set` update commands to the submitter.
- [ ] Preserve current standalone local mutation behavior when no submitter is configured.
- [ ] Fail closed before local mutation for unsupported cluster-mode writes: createIndexes/dropIndexes, replacement/local-read updates, multi updates, upserts, transactions/retryable writes, unsupported writeConcern.
- [ ] Ensure insert/update coalescers cannot bypass the submitter in cluster mode.
- [ ] Map submitter success into Mongo `ok/n/modified/deleted` responses without claiming replica-set semantics.

## Non-Goals And Boundary

- [ ] No real Raft FSM, leader routing, failover, or redirect implementation.
- [ ] No Mongo replica-set personality.
- [ ] No `majority`, `w>1`, `j:true`, or `wtimeout` semantics yet.
- [ ] No `NotWritablePrimary`/redirect correctness beyond fail-closed submitter errors.
- [ ] No production HA or honest Mongo writeConcern claim; that remains #3044-8.

## Required Tests

- [ ] Fake submitter receives deterministic `insert_batch` for Mongo insert, with collection ref, BSON document format, ids/docs, idempotency key, and expected catalog version.
- [ ] Fake submitter receives deterministic `delete_batch` for `_id` equality deletes.
- [ ] With #3254, fake submitter receives deterministic `update_bson_set` for single `_id` top-level `$set`.
- [ ] Unsupported update shapes reject before submit and before local mutation.
- [ ] Coalesced insert/update paths do not bypass submitter in cluster mode.
- [ ] Submitter errors do not mutate local collections.
- [ ] Standalone insert/update/delete behavior remains unchanged.

## Benchmarks And Evidence

Mongo gateway write routing is performance-sensitive. Include focused before/after evidence if this changes standalone hot paths; otherwise document that cluster-only branches are not active in standalone mode.

## Branch And PR Policy

- Use a topic branch from current `main` after dependencies merge, or a clearly marked synthetic integration branch before then.
- A synthetic PR may be opened for review, but it must not merge before #3249 and, if `$set` is included, #3254 merge and the branch is rebased/revalidated on latest `main`.
- Request Codex, Copilot, and CodeRabbit reviews only after mature code, tests, and evidence.
- Latest-head CI is required before merge.

## Completion Criteria

- Cluster-mode Mongo writes have a stable deterministic submitter seam.
- Cluster mode never silently falls back to local direct mutation for supported or unsupported writes.
- The PR does not claim Mongo primary/writeConcern production semantics.

### PR DESCRIPTION
**Pull request by snissn**
## Objective

Route Mongo gateway cluster-mode mutations through the deterministic native-wire `ClusterSubmitter` seam instead of mutating local collections directly.

Closes #3258.
Refs #3044.

## Context

This is the Mongo gateway submitter adapter for the #3044 single-group HA graph. The lower nativewire submitter seam and `update_bson_set` command support are already merged on `main`, so this branch is based on the current post-#3254 head.

## Non-goals

- No real Raft leader routing, failover, redirect, or replica-set personality.
- No Mongo writeConcern implementation for `majority`, `w>1`, `j:true`, or `wtimeout`.
- No cluster support for index DDL; cluster mode rejects `createIndexes` and `dropIndexes` before local mutation.
- No fallback to local writes when a cluster submitter is configured.

## Design

- Adds `Server.ClusterSubmitter` and `StandaloneOptions.ClusterSubmitter` wiring for Mongo gateway cluster mode.
- Adds a `ClusterCatalogVersionProvider`; standalone cluster mode defaults it to the backend `State().CommitSeq` guard.
- Threads request context through command dispatch so submitter calls observe connection/request cancellation.
- Builds deterministic native-wire entries for `create`, `insert`, `_id` equality delete, and top-level BSON `$set` update.
- Requires submitter response metadata counts for write responses and cross-checks `actual_ack_policy` against the submit result.
- Fails closed for unsupported cluster mutations and writeConcern before local mutation.

## Scope

- Includes:
  - `TreeDB/mongo_gateway/cluster_submitter.go`
  - `TreeDB/mongo_gateway/commands.go`
  - `TreeDB/mongo_gateway/server_core.go`
  - `TreeDB/mongo_gateway/standalone.go`
  - focused Mongo gateway cluster submitter tests
- Excludes:
  - Raft FSM execution
  - read routing/snapshots/recovery
  - multi-group routing
  - production Mongo replica-set semantics

## Correctness

Tests run on latest local head `ff55371e3`:

- `git diff --check`
- `GOWORK=off go test ./TreeDB/mongo_gateway -run 'ClusterSubmitter|BufferedMessage|ServeOne' -count=1`
- `GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry -run 'Cluster|Deterministic|Insert|Delete|UpdateBSONSet|Ack' -count=1`
- `GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire -count=1`
- `GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry ./TreeDB/collections -count=1`

Covered edge cases:

- deterministic command entry routing for insert, create, update_bson_set, and delete
- catalog version section sourced from configured provider
- missing catalog provider rejects before submit
- unsupported update shapes reject before submit and before local mutation
- `createIndexes` / `dropIndexes` reject before submit and before local mutation
- canceled request context reaches the submitter
- missing `inserted_count` fails closed
- submit result `actual_ack_policy` mismatch fails closed

## Performance

No benchmarks run. This PR adds cluster-mode branches that are inactive in standalone local mutation mode unless a submitter is configured; the affected local hot-path packages were validated with the package tests above. No format or storage layout change is included.

## Rollout / Toggle Plan

Default setting: unchanged standalone local mutation path unless `ClusterSubmitter` is configured.

Disable quickly: do not configure `ClusterSubmitter`; the gateway keeps existing standalone behavior.

## Follow-ups

- #3257: raft FSM single-group apply loop on the submitter output.
- #3044: wire the single-group HA endpoint through the accepted slices.
- #3045/#3046: read/snapshot/recovery and multi-group routing after the single-group path is stable.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
Latest-head update for ba600f686:

Scope since prior review:
- Fixed stale catalog guard fallback by requiring an explicit ClusterCatalogVersion provider in cluster-submit mode.
- Rejected AckRaftCommitted for Mongo AckVisible requests unless local visibility is actually proven.
- Preserved ordered update semantics by submitting earlier valid update_bson_set items before a later unsupported update shape returns an error.
- Deduplicated Mongo delete IDs before delete_batch submission.
- Preserved first-insert auto-create semantics in cluster mode by submitting create_collection before insert_batch when the local catalog has no collection.
- Added a per-server cluster idempotency nonce to generated Mongo gateway idempotency keys so process-local sequence numbers do not reuse the same key namespace after restart.
- Preserved earlier ordered deletes before a later delete parse/shape error.
- Merged current origin/main fa5f51628 and resolved the nativewire unsupported-command test conflict in favor of the current CreateIndex guard.

Local validation on ba600f686:
- git diff --check
- GOWORK=off go test ./TreeDB/nativewire -run 'TestClusterSubmitterUnsupportedCommandFailsBeforeMutation|TestClusterSubmitter' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway -run 'TestClusterSubmitter|TestNormalizeStandaloneOptions' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway -run 'ClusterSubmitter|BufferedMessage|ServeOne' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry -run 'Cluster|Deterministic|Insert|Delete|UpdateBSONSet|Ack|Standalone' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry ./TreeDB/collections -count=1

CI/review status:
- Latest-head CI is running on ba600f686.
- Stale non-head runs for superseded heads were canceled where still queued/in progress.
- CodeRabbit auto-triggered on this head but remains rate-limited; current CodeRabbit status context is success/no review findings available.
- Requesting fresh Codex and Copilot review on this mature head next.

**Comment by snissn** [comment]
Latest-head update for ba600f686:

- Merged current main fa5f51628 into the PR branch and resolved the nativewire cluster-submit test conflict.
- Addressed the Codex review findings: explicit catalog-version provider, no AckRaftCommitted-for-visible shortcut, ordered update/delete semantics, duplicate delete ID handling, first-insert auto-create, and restart-safe Mongo gateway idempotency key namespace via per-server nonce.
- Local validation passed on ba600f686: git diff --check; focused nativewire cluster submitter test; focused Mongo cluster submitter tests; gateway/nativewire deterministic command matrix; full mongo_gateway/nativewire packages; wider mongo_gateway/nativewire/internal/raftentry/collections package set.
- Latest-head CI is running on ba600f686. Stale non-head queued/in-progress runs were canceled. CodeRabbit auto-triggered but remains rate-limited, so no fresh findings are available from it yet.

Requesting fresh Codex and Copilot review on this mature head now.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
Latest-head update for 8ed4158ff:

- Addressed the latest Codex P2 findings from ea22a225: cluster-mode update/delete now preserve Mongo missing-collection no-op semantics by returning zero counts without submitting a Raft mutation when the local collection manager reports the namespace absent.
- Added regression coverage for missing-collection update/delete so no submitter calls are made and `{n: 0}` / `{n: 0, nModified: 0}` is returned.
- Stale CI for superseded head ba600f686 was canceled; latest-head CI is now running on 8ed4158ff.
- Requested fresh Codex and Copilot review on 8ed4158ff. CodeRabbit remains rate-limited, so no fresh CodeRabbit review findings are available yet.

Local validation on 8ed4158ff:
- git diff --check
- GOWORK=off go test ./TreeDB/mongo_gateway -run 'TestClusterSubmitter(UpdateMissingCollectionReturnsZeroCounts|DeleteMissingCollectionReturnsZeroCount|UpdateSubmitsPriorOrderedItemsBeforeUnsupported|DeleteSubmitsPriorOrderedItemsBeforeUnsupported|DeleteDeduplicatesDuplicateIDs|UpdateBSONSetRoutesCountsAndNoLocalMutation|DeleteRoutesCommandEntry|InsertRoutesCommandEntryNoLocalMutation)' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway -run 'TestClusterSubmitter|TestNormalizeStandaloneOptions' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry -run 'Cluster|Deterministic|Insert|Delete|UpdateBSONSet|Ack|Standalone' -count=1
- GOWORK=off go test ./TreeDB/mongo_gateway ./TreeDB/nativewire ./TreeDB/internal/nativewire ./TreeDB/internal/raftentry ./TreeDB/collections -count=1

File: TreeDB/mongo_gateway/standalone.go [hunk 1]
```diff
@@ -144,6 +147,17 @@ func OpenStandaloneServer(opts StandaloneOptions) (*StandaloneServer, error) {
...
 	server.DefaultIndexStoragePolicy = normalized.DefaultIndexStoragePolicy
+	server.ClusterSubmitter = normalized.ClusterSubmitter
+	server.ClusterCatalogVersion = normalized.ClusterCatalogVersion
+	if server.ClusterCatalogVersion == nil && server.ClusterSubmitter != nil {
+		server.ClusterCatalogVersion = func(context.Context) (uint64, error) {
+			state := backend.State()
+			if state == nil {
+				return 0, errors.New("mongo gateway standalone: backend state is unavailable")
+			}
+			return state.CommitSeq, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/standalone.go]: Fixed: cluster submit mode now requires an explicit ClusterCatalogVersion provider and no longer falls back to CommitSeq.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+		return commandError(commandCodeFailedToParse, "FailedToParse", err.Error())
+	}
+	parsed := make([]mongoUpdateItem, 0, len(updates))
+	for i, update := range updates {
+		item, err := parseMongoUpdateItem(i, update)
+		if err != nil {
+			return mongoUpdateParseCommandError(err)
+		}
+		if !item.bsonSetFieldsOK {
+			return commandError(commandCodeBadValue, "BadValue", fmt.Sprintf("updates[%d]: cluster Mongo gateway currently supports only top-level BSON $set updateOne by _id", i))
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:139]: Fixed: ordered update parsing now submits prior valid update_bson_set items before returning a later unsupported update error.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+		} else if limit != 0 && limit != 1 {
+			return commandError(commandCodeBadValue, "BadValue", "Mongo gateway delete limit must be 0 or 1")
+		}
+		key, err := encodePrimaryKey(id)
+		if err != nil {
+			return commandError(commandCodeBadValue, "BadValue", fmt.Sprintf("deletes[%d]: %v", i, err))
+		}
+		ids = append(ids, key)
+	}
+	deleted, err := s.submitClusterDelete(ctx, name, ids)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:227]: Fixed: cluster deletes deduplicate duplicate encoded Mongo _id values before delete_batch submission.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+	documents, err := commandDocuments(command, sequences, "documents")
+	if err != nil {
+		return commandError(commandCodeFailedToParse, "FailedToParse", err.Error())
+	}
+	format := s.clusterCollectionFormat(name)
+	ids, stored, err := prepareInsertDocuments(documents, format)
+	if err != nil {
+		return commandError(commandCodeBadValue, "BadValue", err.Error())
+	}
+	inserted, err := s.submitClusterInsert(ctx, name, format, ids, stored)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:95]: Fixed: first insert into an absent collection now submits create_collection before insert_batch in cluster mode.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+	}
+	return sections, seq
+}
+
+func (s *Server) clusterMutationSections(command iwire.CommandID, commandName, collection string, catalogVersion uint64, payload ...iwire.Section) ([]iwire.Section, uint64) {
+	seq := s.nextClusterSubmit.Add(1)
+	sections := make([]iwire.Section, 0, 5+len(payload))
+	sections = append(sections,
+		iwire.Section{ID: iwire.SectionCommandHeader, Bytes: iwire.AppendCommandHeader(nil, iwire.CommandHeader{ID: command, Version: 1})},
+		iwire.Section{ID: iwire.SectionIdempotencyKey, Bytes: []byte("mongo-gateway/" + commandName + "/" + strconv.FormatUint(seq, 10))},
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go]: Fixed: generated cluster idempotency keys include a per-server nonce, avoiding reuse across gateway restarts.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+		} else if limit != 0 && limit != 1 {
+			return commandError(commandCodeBadValue, "BadValue", "Mongo gateway delete limit must be 0 or 1")
+		}
+		key, err := encodePrimaryKey(id)
+		if err != nil {
+			return commandError(commandCodeBadValue, "BadValue", fmt.Sprintf("deletes[%d]: %v", i, err))
+		}
+		ids = append(ids, key)
+	}
+	deleted, err := s.submitClusterDelete(ctx, name, ids)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:227]: Fixed: ordered deletes now submit already parsed valid IDs before returning a later parse or shape error.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 7]
```diff
@@ -0,0 +1,667 @@
...
+	var matched, modified int32
+	for i, update := range updates {
+		item, err := parseMongoUpdateItem(i, update)
+		if err != nil {
+			return mongoUpdateParseCommandError(err)
+		}
+		if !item.bsonSetFieldsOK {
+			return commandError(commandCodeBadValue, "BadValue", fmt.Sprintf("updates[%d]: cluster Mongo gateway currently supports only top-level BSON $set updateOne by _id", i))
+		}
+		matchedOne, modifiedOne, err := s.submitClusterUpdateBSONSet(ctx, name, item)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:141]: Fixed in 8ed4158ff: update against a locally missing collection returns zero counts without submitting update_bson_set.

File: TreeDB/mongo_gateway/cluster_submitter.go [hunk 8]
```diff
@@ -0,0 +1,667 @@
...
+			return commandError(commandCodeBadValue, "BadValue", fmt.Sprintf("deletes[%d]: %v", i, err))
+		}
+		keyString := string(key)
+		if _, ok := seenIDs[keyString]; ok {
+			continue
+		}
+		seenIDs[keyString] = struct{}{}
+		ids = append(ids, key)
+	}
+	deleted, err := s.submitClusterDelete(ctx, name, ids)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/mongo_gateway/cluster_submitter.go:227]: Fixed in 8ed4158ff: delete against a locally missing collection returns n: 0 without submitting delete_batch.
